### ISSUE TITLE
Attach VSCode debugger

### ISSUE THREAD

**Issue by maxhor** [Microsoft/vscode-chrome-debug#131]
I have VSCode debug working if I 'launch' the app. (the app is just the angular.io 'hero' tutorial) However: launching is not a serious option as it does not allow to have any other web page open in Chrome.  
'Attach' is therefore the better option. But I just don't understand how that works. Documentation says:

`You must launch Chrome with remote debugging enabled in order for the extension to attach to it.`

So I start Chrome with. 

`"path/chrome.exe" --remote-debugging-port=9222" `

I do not really understand what I am doing, but I guess I have to type in http://localhost:9222/ in  Chromes address bar

Then when launching the app in VSCode, I get the following errors:

 ```
 ›Warning: Found more than one valid target page. Attaching to the first one. Available pages: ["http://localhost:9222/","https://github.com/Microsoft/vscode-chrome-debug","https://github.com/Microsoft/vscode-chrome-debug/issues/new","https://github.com/Microsoft/vscode-chrome-debug"]
Not allowed to load local resource: chrome://extension-icon/cfhdojbkjhnklbpkdaibdccddilifddb/24/1
Not allowed to load local resource: chrome://extension-icon/jpfbieopdmepaolggioebjmedmclkbap/24/1
Failed to load resource: the server responded with a status of 404 (Not Found) (http://localhost:9222/undefined)
```

When I switch 'diagnostic logging' to true and start debug again, there is no error. But nothing happens in the debugger. The logging most interesting messages are:

```
›Attempting to attach on port 9222
  ›Warning: Found more than one valid target page. Attaching to the first one. Available pages: ["https://github.com/Microsoft/vscode-chrome-debug/issues/new","http://localhost:9222/","https://github.com/Microsoft/vscode-chrome-debug","https://github.com/Microsoft/vscode-chrome-debug"]
```

and 

`›Paths.scriptParsed: could not resolve https://assets-cdn.github.com/assets/frameworks-20e2831691c9bb0a4dc1bd778529fc1c16ec8c8a24b32d9964f984772d2eb24b.js to a file in the workspace. webRoot: h:\VSCode\plunk/app`

and 

```
›From client: setBreakpoints({"source":{"path":"h:\\VSCode\\plunk\\app\\app.component.ts"},"lines":[41,40,30,31,45,35],"breakpoints":[{"line":41},{"line":40},{"line":30},{"line":31},{"line":45},{"line":35}]})
  ›Paths.setBP: No target url cached for client path: h:\VSCode\plunk\app\app.component.ts, waiting for target script to be loaded.
```

My launch.json: 

```
{
			"name": "Attach",
			"type": "chrome",
			"request": "attach",
                        "diagnosticLogging": false,  // and true   
			"port": 9222
		}

```

By the way: if I start Chrome and go to localhost:9222: will that invoke the IIS server on my Windows10 machine? (and not the 'lite' server) Any way:

I would very much appreciate some guiding and readingadvice: 

- to understand what I am doing when attaching visual code to Chrome

- get it going

Tx., Marcel

**Comment by auchenberg** [comment]
Hi Marcel,

Can you try to update the attach config to this (`webRoot` added)
```json
        {
            "name": "Chrome Attach",
            "type": "chrome",
            "request": "attach",
            "port": 9222,
            "sourceMaps": true,
            "webRoot": "${workspaceRoot}",
            "diagnosticLogging": false
        }
```

**Comment by maxhor** [comment]
I have been experimenting with "webRoot" before making my post. It makes no difference. So with launch.json 

```
{
            "name": "Chrome Attach",
            "type": "chrome",
            "request": "attach",
            "port": 9222,
            "sourceMaps": true,
            "webRoot": "${workspaceRoot}/app",
            "diagnosticLogging": false
        }

```
I still get: 

```
Failed to load resource: the server responded with a status of 404 (Not Found) (http://localhost:9222/favicon.ico)
Not allowed to load local resource: chrome://extension-icon/cfhdojbkjhnklbpkdaibdccddilifddb/24/1
Not allowed to load local resource: chrome://extension-icon/jpfbieopdmepaolggioebjmedmclkbap/24/1
Failed to load resource: the server responded with a status of 404 (Not Found) (http://localhost:9222/undefined)
```

My files reside in the 'app' folder under the project root, so I used

   `"webRoot": "${workspaceRoot}/app"`, 

But for the sake of it, I also tried the setting you literally gave: 

 `"webRoot": "${workspaceRoot}/app"`

**Comment by auchenberg** [comment]
I see. Based on the logging it doesn't seem like you have a tab open with your local server?

Try to have **one** tab open in Chrome visiting your local server ala `http://localhost:8080`, and then connect from VS Code?

**Comment by maxhor** [comment]
I did open exactly one tab. 

In case I do not open a tab in Chrome at all, I get another message. That message comes on top of the screen. 

`[webkit-debug-adapter] Cannot connect to the target: connect ECONNREFUSED 127.0.0.1:9222`

In that case, with diagnostic logging off: there is no message in the console 

I go to localhost:9222 (not localhost:8080)

**Comment by roblourens** [comment]
In your original launch config, you don't have sourcemaps enabled - try adding `"sourcemaps": true`. Also, the errors you see about the favicon and extensions are fine. And the errors you posted are from vscode connecting to a "localhost:9222" so make sure that you don't have another instance of chrome hiding somewhere.

**Comment by maxhor** [comment]
In my latest launch.json, I had sourcemaps enabled. So there is not the problem. 

But doing some more thinking, I realise now, that I do not understand how to make the connection between the project in vscode and the Chrome page.

Documentation says: 

> Launch Chrome and navigate to **your** page.

What page is that? Until now, I just typed the following address in the browser: 

`localhost:9222`

But that is not ok I think now. 

Should I set up a website in IIS ? (My workstation is Windows10. ) That website pointing to the rootdirectory of my project in VSCode? In Chrome connect to the index.html from that website? And then start debug in my vscode project?

**Comment by roblourens** [comment]
Up to you - you'll either want a server set up, or you can load your page like `file:///c:/path/to/index.html`. If you've loaded your page somehow before, that's the url you want to use. Not `localhost:9222`, that's a page that chrome provides for debugging.

**Comment by maxhor** [comment]
I got it to work finally. I set up the website in IIS, bind in my case to port 8081. Start website in IIS. And then I start Chrome with (ofcourse)
`path\chrome.exe --remote-debugging-port=9222`
Then navigate to localhost:8081, website shows in the browser. And then, in vscode, I am able to attach.
So very nice. Thank you. 

By the way: typing in the browser: 
 `file:///c:/path/to/index.html does not work`

does not work. The raw content of index.html shows, but Angular is not fired. F12 in Chrome tells me:

> system.src.js:1085 XMLHttpRequest cannot load file:///H:/VSCode/plunk/app/main.js. Cross origin requests are only supported for protocol schemes: http, data, chrome, chrome-extension, https, chrome-extension-resource.

There is a post about this on stackoverflow: the website must be opened in the server to prevent a cross origin problem.

> http://stackoverflow.com/questions/27742070/angularjs-error-cross-origin-requests-are-only-supported-for-protocol-schemes
